hax TypeError: Object of type StobIoqError is not JSON serializable on disk failure

{noformat}
Oct 12 19:27:44 sage-tier3-1 kernel: sd 0:0:12:0: [sdl] tag#17 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE cmd_age=4s
Oct 12 19:27:44 sage-tier3-1 kernel: sd 0:0:12:0: [sdl] tag#17 Sense Key : Medium Error [current] [descriptor] 
Oct 12 19:27:44 sage-tier3-1 kernel: sd 0:0:12:0: [sdl] tag#17 Add. Sense: Unrecovered read error
Oct 12 19:27:44 sage-tier3-1 kernel: sd 0:0:12:0: [sdl] tag#17 CDB: Read(32)
Oct 12 19:27:44 sage-tier3-1 kernel: sd 0:0:12:0: [sdl] tag#17 CDB[00]: 7f 00 00 00 00 00 00 18 00 09 20 00 00 00 00 00
Oct 12 19:27:44 sage-tier3-1 kernel: sd 0:0:12:0: [sdl] tag#17 CDB[10]: 00 00 5c 80 00 00 5c 80 00 00 00 00 00 00 00 80
Oct 12 19:27:44 sage-tier3-1 kernel: blk_update_request: critical medium error, dev sdl, sector 190296
Oct 12 19:27:44 sage-tier3-1 motr-server[2385]: motr[02385]:  f830   WARN  [stob/ioq.c:443:ioq_io_error]  IO error: stob_id={<100000000000000:adf11e>,<100000000000000:3f>} conf_sdev=<6400000000000001:111>
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: motr[01709]:  2800   WARN  [ha/halon/interface.c:409:halon_interface_msg_received_cb]  hl=0x7f8f50044070 ep=172.18.1.7@o2ib:12345:2:2 epoch=0 tag=80 type=1
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: motr[01709]:  c670   WARN  [hax.c:433:handle_stob_ioq]  Got STOB_IOQ
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: 2021-10-12 19:27:44,979 [INFO] {Dummy-1} fid=0x6400000000000001:0x111, sie_conf_sdev=0x6400000000000001:0x111
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: motr[01709]:  c5e0   WARN  [ha/halon/interface.c:1135:m0_halon_interface_delivered]  hl=0x7f8f50044070 ep=172.18.1.7@o2ib:12345:2:2 epoch=0 tag=80 type=1
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: 2021-10-12 19:27:44,980 [DEBUG] {qconsumer-1} Got StobIoqError(group=81597, fid=0x6400000000000001:0x111, conf_sdev=0x6400000000000001:0x111, stob_id=StobId(domain_fid=0x100000000000000:0xadf11e, fid=0x100000000000000:0x3f), fd=46, opcode=1, rc=18446744073709551611, offset=96468992, size=1048576, bshift=12) message from planner
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: 2021-10-12 19:27:44,980 [INFO] {qconsumer-1} Stob IOQ: 0x6400000000000001:0x111
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: 2021-10-12 19:27:44,981 [ERROR] {qconsumer-1} **ERROR**
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: Traceback (most recent call last):
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/handler.py", line 227, in _do_work
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: payload = dump_json(item)
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 1492, in dump_json
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: return simplejson.dumps(obj, for_json=True)
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/simplejson/__init__.py", line 412, in dumps
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: **kw).encode(obj)
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/simplejson/encoder.py", line 296, in encode
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: chunks = self.iterencode(o, _one_shot=True)
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/simplejson/encoder.py", line 378, in iterencode
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: return _iterencode(o, 0)
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/simplejson/encoder.py", line 273, in default
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: o.__class__.__name__)
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: TypeError: Object of type StobIoqError is not JSON serializable
Oct 12 19:27:44 sage-tier3-1 hare-hax[1709]: 2021-10-12 19:27:44,982 [DEBUG] {qconsumer-1} Waiting for the next message
{noformat}

--- a/hax/hax/message.py
+++ b/hax/hax/message.py
@@ -17,7 +17,7 @@
 #
 
 import queue as q
-from dataclasses import dataclass, field
+from dataclasses import dataclass, field, fields
 from queue import Queue
 from typing import Any, List, Optional
 
@@ -139,6 +139,37 @@ class StobIoqError(BaseMessage):
     offset: int
     size: int
     bshift: int
+
+    def for_json(self):
+        parts = {}
+
+        def as_str(a):
+            return str(a)
+
+        def as_repr(a):
+            return repr(a)
+
+        def as_is(a):
+            return a
+
+        for fld in fields(self):
+            f_type = fld.type
+            f_name = fld.name
+            if f_name == 'group':
+                # group is a thing used by WorkPlanner for task scheduling
+                # logic. We don't need to expose it
+                continue
+            val = getattr(self, f_name)
+            to_str = as_str
+            if f_type is Fid:
+                to_str = as_repr
+            elif f_type is StobId:
+                to_str = as_is
+            elif f_type is int or f_type is None:
+                to_str = as_is
+            parts[fld.name] = to_str(val)
+
+        return parts
 
 
 class Die(BaseMessage):

--- a/hax/test/integration/test_motr.py
+++ b/hax/test/integration/test_motr.py
@@ -22,15 +22,16 @@ import logging
 from typing import Any, Callable
 
 import pytest
+
 from hax.handler import ConsumerThread
 from hax.message import (BaseMessage, Die, FirstEntrypointRequest,
-                         HaNvecGetEvent)
+                         HaNvecGetEvent, StobId, StobIoqError)
 from hax.motr import Motr, WorkPlanner
 from hax.motr.delivery import DeliveryHerald
 from hax.motr.ffi import HaxFFI
-from hax.types import (Fid, FidStruct, HaNoteStruct, HAState, Profile, HaNote,
+from hax.types import (Fid, FidStruct, HaNote, HaNoteStruct, HAState, Profile,
                        ServiceHealth, Uint128)
-from hax.util import create_process_fid, create_profile_fid
+from hax.util import create_process_fid, create_profile_fid, dump_json
 
 from .testutils import (AssertionPlan, FakeFFI, Invocation, TraceMatcher,
                         tr_and, tr_method, tr_not)
@@ -942,3 +943,23 @@ def test_broadcast_io_service_failure(mocker, planner, motr, consumer,
     assert AssertionPlan(tr_and(tr_method('ha_broadcast'),
                                 node_fid_failed())).not_exists(traces), \
         'Node failure should not be broadcast'
+
+
+# flake8: noqa
+
+
+def test_bq_stob_serializeable():
+    stob = StobId(Fid(12, 13), Fid(14, 15))
+    msg = StobIoqError(fid=Fid(5, 6),
+                       conf_sdev=Fid(1, 2),
+                       stob_id=stob,
+                       fd=42,
+                       opcode=4,
+                       rc=2,
+                       offset=0xBF,
+                       size=100,
+                       bshift=4)
+    first = dump_json(stob)
+    assert first == '{"domain_fid": "0xc:0xd", "fid": "0xe:0xf"}'
+    second = dump_json(msg)
+    assert second == '{"fid": "0x5:0x6", "conf_sdev": "0x1:0x2", "stob_id": {"domain_fid": "0xc:0xd", "fid": "0xe:0xf"}, "fd": 42, "opcode": 4, "rc": 2, "offset": 191, "size": 100, "bshift": 4}'

--- a/hax/test/integration/test_server.py
+++ b/hax/test/integration/test_server.py
@@ -22,11 +22,12 @@ from typing import List
 
 import pytest
 import simplejson
-from hax.message import BroadcastHAStates
+from hax.message import BroadcastHAStates, StobId, StobIoqError
 from hax.motr import WorkPlanner
 from hax.motr.delivery import DeliveryHerald
 from hax.server import ServerRunner
 from hax.types import Fid, HAState, MessageId, ServiceHealth
+from hax.util import dump_json
 
 
 @pytest.fixture
@@ -152,3 +153,56 @@ async def test_bq_stob_message_type_recognized(hax_client, planner, herald,
     planner.add_command.assert_called_once_with(
         ContainsStates(
             [HAState(fid=Fid(0x1, 0x4), status=ServiceHealth.FAILED)]))
+
+
+async def test_bq_stob_message_deserialized(hax_client, planner, herald,
+                                            consul_util, mocker):
+    def fake_get(key):
+        ret = {'bq-delivered/192.168.0.28': ''}
+        return ret[key]
+
+    mocker.patch.object(herald, 'wait_for_any')
+    #
+    # InboxFilter will try to read epoch - let's mock KV operations
+    stob = StobId(Fid(12, 13), Fid(14, 15))
+    msg = StobIoqError(fid=Fid(5, 6),
+                       conf_sdev=Fid(0x103, 0x204),
+                       stob_id=stob,
+                       fd=42,
+                       opcode=4,
+                       rc=2,
+                       offset=0xBF,
+                       size=100,
+                       bshift=4)
+
+    # Here we make sure that rea StobIoqError can be used as the payload
+    # for STOB_IOQ_ERROR bq message.
+    stob_payload = dump_json(msg)
+    parsed_stob = simplejson.loads(stob_payload)
+
+    mocker.patch.object(consul_util.kv, 'kv_put')
+    mocker.patch.object(consul_util.kv, 'kv_get', fake_get)
+    event_payload = {'message_type': 'STOB_IOQ_ERROR', 'payload': parsed_stob}
+    event_str = simplejson.dumps(event_payload)
+    b64: bytes = b64encode(event_str.encode())
+    b64_str = b64.decode()
+
+    payload = [{
+        'Key': 'bq/12',
+        'CreateIndex': 1793,
+        'ModifyIndex': 1793,
+        'LockIndex': 0,
+        'Flags': 0,
+        'Value': b64_str,
+        'Session': ''
+    }]
+    # Test execution
+    resp = await hax_client.post('/watcher/bq', json=payload)
+    # Validate now
+    if resp.status != 200:
+        resp_json = await resp.json()
+        logging.getLogger('hax').debug('Response: %s', resp_json)
+    assert resp.status == 200
+    planner.add_command.assert_called_once_with(
+        ContainsStates(
+            [HAState(fid=Fid(0x103, 0x204), status=ServiceHealth.FAILED)]))

